cmd/process-agent/collector.go

```diff
@@ -92,24 +97,83 @@ func (l *Collector) runCheck(c checks.Check, payloads chan checkPayload) {
 	messages, err := c.Run(l.cfg, atomic.AddInt32(&l.groupID, 1))
 	if err != nil {
 		log.Errorf("Unable to run check '%s': %s", c.Name(), err)
-	} else {
-		payloads <- checkPayload{messages, c.Name()}
-		// update proc and container count for info
-		updateProcContainerCount(messages)
-		if !c.RealTime() {
-			d := time.Since(s)
-			switch {
-			case runCounter < 5:
-				log.Infof("Finished %s check #%d in %s", c.Name(), runCounter, d)
-			case runCounter == 5:
-				log.Infof("Finished %s check #%d in %s. First 5 check runs finished, next runs will be logged every 20 runs.", c.Name(), runCounter, d)
-			case runCounter%20 == 0:
-				log.Infof("Finish %s check #%d in %s", c.Name(), runCounter, d)
-			}
+		return
+	}
+
+	for _, m := range messages {
+		body, err := encodePayload(m)
+		if err != nil {
+			log.Errorf("Unable to encode message: %s", err)
+			continue
+		}
+
+		extraHeaders := make(http.Header)
+		extraHeaders.Set(api.HostHeader, l.cfg.HostName)
+		extraHeaders.Set(api.ProcessVersionHeader, Version)
+		extraHeaders.Set(api.ContainerCountHeader, strconv.Itoa(getContainerCount(m)))
+
+		if cid, err := clustername.GetClusterID(); err == nil && cid != "" {
+			extraHeaders.Set(api.ClusterIDHeader, cid)
+		}
+
+		if err := l.acquireQueueBytes(len(body)); err != nil {
+			log.Errorf("Unable to enqueue message: %s", err)
+			continue
+		}
+
+		payload := checkPayload{
+			name:    c.Name(),
+			body:    body,
+			headers: extraHeaders,
+		}
+		select {
+		case payloads <- payload:
+		default:
+			bytesEnqueued := atomic.LoadInt32(&l.bytesEnqueued)
+			log.Warnf("Delivery queue is full.  Size=%d, Bytes Enqueued=%d.  Waiting for space...", len(payloads), bytesEnqueued)
+
+			payloads <- payload
+		}
+	}
+
+	// update proc and container count for info
+	updateProcContainerCount(messages)
+	if !c.RealTime() {
+		d := time.Since(s)
+		switch {
+		case runCounter < 5:
+			log.Infof("Finished %s check #%d in %s", c.Name(), runCounter, d)
+		case runCounter == 5:
+			log.Infof("Finished %s check #%d in %s. First 5 check runs finished, next runs will be logged every 20 runs.", c.Name(), runCounter, d)
+		case runCounter%20 == 0:
+			log.Infof("Finish %s check #%d in %s", c.Name(), runCounter, d)
 		}
 	}
 }
 
+func (l *Collector) acquireQueueBytes(bytesRequested int) error {
+	max := int32(l.cfg.QueueBytes)
+
+	for {
+		currentSize := atomic.LoadInt32(&l.bytesEnqueued)
+		newSize := int32(bytesRequested) + currentSize
+		if newSize > max {
+			return fmt.Errorf("cannot acquire %d bytes of queue space, queue size %d would exceed capacity %d", bytesRequested, newSize, max)
+		}
+		if atomic.CompareAndSwapInt32(&l.bytesEnqueued, currentSize, newSize) {
+			break
+		}
+	}
```

Comment:
Does this need to be a busy loop? Is there any downside re: burning cycles if the scheduler doesn't preempt the goroutine?

Reply:
In most cases this should succeed immediately, the only time a goroutine would have to spin s if N checks concurrently attempt to acquire queue space at the same time.  Only one check will succeed forcing the others to retry again.

This could also be rewritten with a lock if you prefer.